ICESat-2 ML tutorial - photon classification on ATL07 sea ice data

weiji14 opened the pull request:
Draft tutorial for doing ICESat-2 ATL07 photon classification into 3 surface types (open water, thin ice, thick/snow-covered ice). Deciding to do a reimplementation of @YoungHyunKoo's code at https://github.com/YoungHyunKoo/IS2_ML.

**Preview** at https://deploy-preview-17--icesat2-website2024.netlify.app/tutorials/machine-learning/photon_classifier

![ATL07_point_cloud_classifier](https://github.com/user-attachments/assets/d61521a4-d27a-4eb8-b886-3d92c5516c32)

Excalidraw file: [ATL07_point_cloud_classifier.excalidraw.tar.gz](https://github.com/user-attachments/files/16655914/ATL07_point_cloud_classifier.excalidraw.tar.gz)

TODO:
- [x] Initial layout with learning objectives and sections
- [x] Pre-processing code to read ATL07 sea ice data (6 data variables) from HDF5 to a `geopandas.GeoDataFrame`
- [x] Find coincident Sentinel-2 image to extract surface reflectance values
- [x] Show how to move data from disk/CPU to GPU
- [x] Architect the Machine Learning model
- [ ] Show training results

Xref: https://github.com/uwhackweek/schedule-2024/issues/38

References:
- Koo, Y., Xie, H., Kurtz, N. T., Ackley, S. F., & Wang, W. (2023). Sea ice surface type classification of ICESat-2 ATL07 data by using data-driven machine learning model: Ross Sea, Antarctic as an example. Remote Sensing of Environment, 296, 113726. https://doi.org/10.1016/j.rse.2023.113726

weiji14 committed:
Initial jupytext notebook with outline and ATL07 to geopandas script

First draft with a rough layout of sections for the ICESat-2 ML photon classification tutorial. Included learning objectives, and some initial code to read ATL07 sea ice data from HDF5 to a geopandas.GeoDataFrame. Deciding to do a reimplementation of the Koo et al., 2023 paper with code at https://github.com/YoungHyunKoo/IS2_ML.

weiji14 commented on book/tutorials/photon_classifier.py:
```diff
@@ -0,0 +1,178 @@
+# ## Part 0: Setup
+
+# %%
+import geopandas as gpd
```
Getting this error on the Netlify build:

```python-traceback
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[1], line 1
----> 1 import geopandas as gpd
      2 import h5py
      3 import icepyx as ipx

ModuleNotFoundError: No module named 'geopandas'
```

Might need to either sync this repo's environment.yml file to match with CryoCloud [hub-image](https://github.com/CryoInTheCloud/hub-image), or push the pre-rendered notebook cells.

weiji14 replied:
Resolved in https://github.com/ICESAT-2HackWeek/website-2024/pull/18

weiji14 committed:
Save ATL07 photon data to GeoParquet file with ZSTD compression

Show how to save geopandas.GeoDataFrame to a GeoParquet file, and load it back again. Also put down some notes about compression codecs.

weiji14 committed:
Writeup sub-section on moving data from CPU to GPU

Some quick code to convert the geopandas.GeoDataFrame to a torch.Tensor and put it in a torch DataLoader. Showing how to move data from CPU to GPU using the `.to` method. Might modify this section's title/subtitle later depending on how the code goes.

weiji14 committed:
Add ICESat-2 photon classification tutorial to index table

One more entry in the tutorial index page. Putting down Machine Learning and Pytorch as the topics, and ATL07 as the dataset used for now.

weiji14 committed:
Merge branch 'main' into photon_classifier

weiji14 committed:
Add pytorch (cpu build) to conda environment

weiji14 commented on book/tutorials/photon_classifier.py:
```diff
@@ -0,0 +1,261 @@
+# s3links
+
+# %%
+# Authenticate using NASA EarthData login; enter your user id and password when prompted
+credentials = region_ross.s3login_credentials
+credentials
```
Hmm, do we have some secrets in this repo for Earthdata login credentials?

```python-traceback
---------------------------------------------------------------------------
AuthenticationError                       Traceback (most recent call last)
Cell In[5], line 2
      1 # Authenticate using NASA EarthData login; enter your user id and password when prompted
----> 2 credentials = region_ross.s3login_credentials
      3 credentials

File ~/micromamba/envs/hackweek/lib/python3.11/site-packages/icepyx/core/auth.py:112, in EarthdataAuthMixin.s3login_credentials(self)
    109 # Only generate s3login_credentials the first time credentials are accessed, or if an hour
    110 # has passed since the last login
    111 if self._s3login_credentials is None:
--> 112     set_s3_creds()
    113 elif (datetime.datetime.now() - self._s3_initial_ts) >= datetime.timedelta(
    114     hours=1
    115 ):
    116     set_s3_creds()

File ~/micromamba/envs/hackweek/lib/python3.11/site-packages/icepyx/core/auth.py:106, in EarthdataAuthMixin.s3login_credentials.<locals>.set_s3_creds()
    104 def set_s3_creds():
    105     """Store s3login creds from `auth`and reset the last updated timestamp"""
--> 106     self._s3login_credentials = self.auth.get_s3_credentials(daac="NSIDC")
    107     self._s3_initial_ts = datetime.datetime.now()

File ~/micromamba/envs/hackweek/lib/python3.11/site-packages/icepyx/core/auth.py:76, in EarthdataAuthMixin.auth(self)
     74 # check for a valid auth response
     75 if auth.authenticated is False:
---> 76     raise AuthenticationError(
     77         "Earthdata authentication failed. Check output for error message"
     78     )
     79 else:
     80     self._auth = auth

AuthenticationError: Earthdata authentication failed. Check output for error message
```

I remember asking two years ago and there was a solution at https://github.com/ICESAT-2HackWeek/website2022/pull/40#issuecomment-1072890483 to use non-s3 access. Will need to see what the new syntax is...

weiji14 replied:
Refactored from `icepyx` to `earthaccess` in d616d77b11d745add2bde0a561c1de45b0e2cdc1. Still getting an `AuthenticationError` though, let me open another thread.

weiji14 committed:
Refactor to get ATL07 using earthaccess instead of icepyx

Less boilerplate s3fs code to manage, and not using icepyx means this should run on the Pangeo pytorch-notebook docker image too!

weiji14 commented on book/tutorials/photon_classifier.py:
```diff
@@ -0,0 +1,243 @@
+# TODO: copy Table 1 from Koo et al., 2023 paper
+
+# %%
+# Authenticate using NASA EarthData login
+auth = earthaccess.login()
+s3 = earthaccess.get_s3fs_session(daac="NSIDC")  # Start an AWS S3 session
```
Continuing from https://github.com/ICESAT-2HackWeek/website-2024/pull/17#discussion_r1708123531, this is the current error:

```python-traceback
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[2], line 3
      1 # Authenticate using NASA EarthData login
      2 auth = earthaccess.login()
----> 3 s3 = earthaccess.get_s3fs_session(daac="NSIDC")  # Start an AWS S3 session

File ~/micromamba/envs/hackweek/lib/python3.11/site-packages/earthaccess/api.py:352, in get_s3fs_session(daac, provider, results)
    350         session = earthaccess.__store__.get_s3fs_session(endpoint=endpoint)
    351         return session
--> 352 session = earthaccess.__store__.get_s3fs_session(daac=daac, provider=provider)
    353 return session

AttributeError: 'NoneType' object has no attribute 'get_s3fs_session'
```

Not sure if there's a way to pass auth credentials to GitHub Actions and/or Netlify build so that this line works?

tsutterley replied:
Can definitely add secrets for `EARTHDATA_USERNAME` and `EARTHDATA_PASSWORD`, which would enable `earthdata` to login during actions. However GitHub actions run on azure, so s3 access isn't available. I think we'd have to build a self-hosted runner that deploys on AWS.

scottyhq replied:
Ideally we could execute these notebooks *on a remotely booted CryoCloud user server*. I think all the pieces exist to do this nowadays? @yuvipanda I saw this relevant issue https://github.com/NASA-IMPACT/veda-jupyterhub/issues/46 as I was exploring your latest awesome 2i2c tech including https://github.com/yuvipanda/jupyter-sshd-proxy. Is there a straightforward way to *start* a user server via CI? Then you'd just have to fire up an ssh connection, execute a notebook, copy and commit the rendered version.

weiji14 replied:
Another option is if CryoCloud had a BinderHub, we could follow Project Pythia and use `execute_notebooks: binder` in `_config.yml` (Example [here](https://github.com/ProjectPythia/cesm-lens-aws-cookbook/blob/55fbdc7c03a123bd7736559bf2a300e33861147c/_config.yml#L10), xref https://discourse.pangeo.io/t/statement-of-need-integrating-jupyterbook-and-jupyterhubs-via-ci/2705/14). The ssh option would be pretty cool to get working though!

scottyhq replied:
@weiji14 For now I recommend executing the notebook on CryoCloud, saving with outputs and adding an entry to not execute it in CI here:

https://github.com/ICESAT-2HackWeek/website-2024/blob/3958f0273e208c502343552ffd23f11b73ac4beb/book/_config.yml#L38-L39

weiji14 committed:
Search for Sentinel-2 imagery captured at same time as ATL07 track

Looking for a coincident alignment of two satellites (ICESat-2 and Sentinel-2) capturing data at the same time! Managed to find a coincident capture on 2019-02-24, though haven't checked if the spatial extent matches yet. Can improve the search algorithm later by expanding the search time window (+/- X minutes) and using a more exact bounding box search in the STAC API query.

weiji14 committed:
Second exact spatial intersection search using ATL07 line track

Temporal match wasn't enough, so adding the spatial match as well. Metadata on ICESat-2 was lacking unfortunately, so need to open the ATL07 HDF5 file to get the xy coordinates and build a linestring from it to pass to the STAC query. Managed to find a lucky coincident match on 2019-10-31, and have verified that the crossover is valid.

weiji14 committed:
Add more columns to GeoDataFrame and filter out cloudy points

Add the `x_atc`, `layer_flag` and `height_segment_ssh_flag` data variables to the GeoDataFrame which will be useful for plotting/filtering later. Using `height_segment_ssh_flag` to remove points that might be affected by clouds.

weiji14 committed:
Plot ATL07 tracks on top of Sentinel-2 image

Get the Sentinel-2 RGB image, reproject the ATL07 points and subset to the image's bounding box, then plot them both using PyGMT! The plot colors sea ice points as blue, and sea surface (water) points as orange.

weiji14 committed:
Label surface type of ATL07 points using Sentinel-2 Red band pixel value

Use PyGMT's grdtrack to get the Sentinel-2 Red band's pixel values sampled at every ATL07 xy point, and then apply a simple threshold to classify into water (dark), thin ice (gray) and thick ice (white).

weiji14 committed:
Rename Part 2 to DataLoader and Model architecture

Reorganizing some content so Part 2 is focused on preparing the DataLoader and neural network model architecture. Have now moved the dataloader for-loop to Part 3 'Training' and commented out the to CUDA parts. Also calculated "hist_mean_median_h_diff" column which is the actual variable we want to use in training.

weiji14 committed:
Architect PhotonClassificationModel and writeup ML model choices

Writing up section about choosing a machine learning algorithm, including ML models with different levels of complexity from decision trees to neural networks and state-of-the-art models. Also implemented a simple multi-layer perceptron model based on the description in Koo et al., 2023's paper (but without the tanh activation).

weiji14 committed:
Construct main training loop for ML model

Finally got to the actual neural network model training! Now properly splitting the mini-batch data into input and target tensors, passing the input into the model to get the prediction, and minimizing the loss between prediction and target. Needed to do some ugly dtype casting to prevent `RuntimeError`s. Trying to keep this fairly basic without train/validation splits, and only ran this for 3 epochs. Have shifted some markdown blocks up where they belong too.

weiji14 committed:
Merge branch 'main' into photon_classifier

weiji14 committed:
Add instructions to install pytorch in first code cell

Default CryoCloud docker image won't have Pytorch, so will need to install it at the first step.

weiji14 committed:
Save geoparquet schema version 1.1.0 and reword note on zstd compression

Default CryoCloud image now has Geopandas 1.x, so can save to a non-beta version of GeoParquet schema now.

weiji14 committed:
Merge branch 'main' into photon_classifier

weiji14 commented:
This should be ready for an initial round of reviews. If possible, I'd appreciate some help with the authentication issue at https://github.com/ICESAT-2HackWeek/website-2024/pull/17#discussion_r1708524246 (need to grab both an ATL07 file and also Sentinel-2), but I can also try to sort that out over the weekend.

There are a few other things I'd like to add to the notebook such as more explanation text at the start, and also show what the trained model's predicted ATL07 photon classifications look like, but that can be done in a follow-up PR.

scottyhq commented on book/tutorials/photon_classifier.py:
```diff
@@ -0,0 +1,596 @@
+# function can be used to do this X/Y sampling from a 1-band grid.
+
+# %%
+df_red = pygmt.grdtrack(
+    grid=da_image.sel(band="red").compute(),  # Choose only the Red band
+    points=gdf.get_coordinates(),  # x/y coordinates from ATL07
+    newcolname="red_band_value",
+    interpolation="n",  # nearest neighbour
+)
```
The first time I ran this on cryocloud I got a traceback: 

```pytb
RuntimeError: Error opening 'https://sentinel-cogs.s3.us-west-2.amazonaws.com/sentinel-s2-l2a-cogs/2/C/ND/2019/10/S2B_2CND_20191031_0_L2A/B04.tif': RasterioIOError('No driver registered.')
```
But re-running succeeded... might be an intermittent thing

weiji14 replied:
Yes, I have to re-run this sometimes too. Maybe there's a short-lived token or something.

scottyhq reviewed (approved):
Thanks for this @weiji14! I wont have time for an in-depth review, but to me this looks like a really great tutorial! I wish we could do the direct S3 access in the CI workflow, but I think the easiest thing for now is just to commit a rendered notebook. Please go ahead and merge once you're happy with it

weiji14 committed:
Add overview flowchart to top of notebook and minor edits

Adding an overview diagram of the ATL07 + Sentinel-2 processing pipeline (illustrated using Excalidraw) to the start of the notebook. Made some minor edits to some of the markdown cells to include more references and explanatory text.

weiji14 committed:
Pre-render Jupyter notebook and move files to machine-learning folder

Pushing the photon_classifier Jupyter Notebook with pre-rendered cells that was ran on CryoCloud. Putting the files under a 'machine-learning' folder, to be consistent with the other tutorials using subfolders.

weiji14 commented:
Thanks Scott, I've added an overview diagram at the top, and have pushed the pre-rendered notebook (6.0MB) :see_no_evil:. Will merge this in now, and might work on some extra stuff at the last section (e.g. show the model training results section a bit better) in a follow-up PR if there's time.

tsutterley commented:
I know this PR is closed, but is this a photon classifier or a segment classifier? Photon classifiers label the individual photon events from ATL03 (e.g. YAPC or the land/veg classifier).

weiji14 commented:
> I know this PR is closed, but is this a photon classifier or a segment classifier? Photon classifiers label the individual photon events from ATL03 (e.g. YAPC or the land/veg classifier).

Ah yes, I should probably have called this a point cloud classifier since ATL07 is based on a aggregate of ATL03 points. Let me fix that in a follow-up PR later.

Edit: Updates happening at https://github.com/ICESAT-2HackWeek/website-2024/pull/17
